### DISCUSSION TITLE
Need help trying to figure out what domains/IPs to whitelist for installing packages

### DISCUSSION THREAD

**Discussion by jpluimers** [microsoft/winget-cli#2304] [General] [answered]
One of the PC's here is protected by a [Fritz!Box](https://en.wikipedia.org/wiki/Fritz!Box) router limiting outgoing traffic only to whitelisted domains or IP addresses. The model does not really matter as the filtering of all models work in the same fashion (but the the interested, it is a [7490](https://en.wikipedia.org/wiki/Comparison_of_Fritz!Box_devices)).

On an unprotected PC, this works splendid:

``` text
c:\bin\bin>winget install -e --id WhatsApp.WhatsApp
Found WhatsApp [WhatsApp.WhatsApp] Version 2.2222.12
This application is licensed to you by its owner.
Microsoft is not responsible for, nor does it grant any licenses to, third-party packages.
Downloading https://web.whatsapp.com/desktop/windows/release/x64/WhatsAppSetup.exe
  ██████████████████████████████   145 MB /  145 MB
Successfully verified installer hash
Starting package install...
Successfully installed
```

On the protected PC, I get this:

``` text
C:\bin\bin>winget install -e --id WhatsApp.WhatsApp --verbose-logs
Failed in attempting to update the source: winget
Failed when searching source: winget
An unexpected error occurred while executing the command:
0x8a15000f : Data required by the source is missing

No packages were found among the working sources.
```

From [#1656](https://github.com/microsoft/winget-cli/issues/1656) (thanks @felipecrs and @jedieaston) I figured out where `--verbose-logs` stores logs (`%LOCALAPPDATA%\Packages\Microsoft.DesktopAppInstaller_8wekyb3d8bbwe\LocalState\DiagOutputDir`) and this is the log:

``` text
2022-07-03 11:14:15.728 [CORE] WinGet, version [1.2.11601], activity [{006FBBA2-C5B3-4C91-B402-223D3FA39F25}]
2022-07-03 11:14:15.729 [CORE] OS: Windows.Desktop v10.0.19044.1766
2022-07-03 11:14:15.729 [CORE] Command line Args: winget  install -e --id WhatsApp.WhatsApp --verbose-logs
2022-07-03 11:14:15.729 [CORE] Package: Microsoft.DesktopAppInstaller v1.17.11601.0
2022-07-03 11:14:15.729 [CORE] IsCOMCall:0; Caller: winget-cli
2022-07-03 11:14:15.758 [CLI ] WinGet invoked with arguments: 'install' '-e' '--id' 'WhatsApp.WhatsApp' '--verbose-logs'
2022-07-03 11:14:15.758 [CLI ] Found subcommand: install
2022-07-03 11:14:15.759 [CLI ] Leaf command to execute: root:install
2022-07-03 11:14:15.763 [CORE] Setting action: Get, Type: Secure, Name: admin_settings
2022-07-03 11:14:15.763 [CORE] Admin settings was not found
2022-07-03 11:14:15.763 [CLI ] Executing command: install
2022-07-03 11:14:15.763 [CORE] Setting action: Get, Type: Secure, Name: user_sources
2022-07-03 11:14:15.763 [CORE] Setting action: Get, Type: Standard, Name: sources_metadata
2022-07-03 11:14:15.764 [YAML] Detected UTF-8
2022-07-03 11:14:15.764 [REPO] GetCurrentSourceRefs: Source named 'microsoft.builtin.desktop.frameworks' from origin Default is hidden and is dropped.
2022-07-03 11:14:15.764 [REPO] Default source requested, multiple sources available, adding all to source references.
2022-07-03 11:14:15.764 [REPO] Adding to source references msstore
2022-07-03 11:14:15.764 [REPO] Adding to source references winget
2022-07-03 11:14:15.765 [CORE] Setting action: Get, Type: Secure, Name: user_sources
2022-07-03 11:14:15.765 [CORE] Setting action: Get, Type: Standard, Name: sources_metadata
2022-07-03 11:14:15.765 [YAML] Detected UTF-8
2022-07-03 11:14:15.765 [REPO] Source past auto update time [5 mins]; it has been at least 27613994 mins
2022-07-03 11:14:16.089 [FAIL] WindowsPackageManager.dll!00007FF92ED457C2: ReturnHr(1) tid(1734) 80072EFD     Msg:[winrt::hresult_error: Kan geen verbinding met de server maken.] 

2022-07-03 11:14:16.089 [FAIL] WindowsPackageManager.dll!00007FF92EBF59D7: LogHr(2) tid(1734) 80072EFD 
2022-07-03 11:14:16.090 [FAIL] D:\a\_work\1\s\external\pkg\src\AppInstallerRepositoryCore\RepositorySource.cpp(53)\WindowsPackageManager.dll!00007FF92ED503A1: (caller: 00007FF92EC4AEA6) LogHr(3) tid(1734) 80072EFD     Msg:[winrt::hresult_error: Kan geen verbinding met de server maken.] 

2022-07-03 11:14:16.090 [REPO] Source add/update failed, waiting a bit and retrying: winget
2022-07-03 11:14:18.323 [FAIL] WindowsPackageManager.dll!00007FF92ED457C2: ReturnHr(2) tid(1734) 80072EFD     Msg:[winrt::hresult_error: Kan geen verbinding met de server maken.] 

2022-07-03 11:14:18.323 [FAIL] WindowsPackageManager.dll!00007FF92EBF59D7: LogHr(5) tid(1734) 80072EFD 
2022-07-03 11:14:18.323 [FAIL] D:\a\_work\1\s\external\pkg\src\AppInstallerRepositoryCore\RepositorySource.cpp(516)\WindowsPackageManager.dll!00007FF92ED4FB13: (caller: 00007FF92EB5BA6D) LogHr(6) tid(1734) 80072EFD     Msg:[winrt::hresult_error: Kan geen verbinding met de server maken.] 

2022-07-03 11:14:18.323 [REPO] Failed to update source: winget
2022-07-03 11:14:18.323 [REPO] Multiple sources available, creating aggregated source.
2022-07-03 11:14:18.323 [REPO] Adding to aggregated source: msstore
2022-07-03 11:14:18.323 [REPO] Custom header not found.
2022-07-03 11:14:18.323 [REPO] Sending http GET request to: https://storeedgefd.dsx.mp.microsoft.com/v9.0/information
2022-07-03 11:14:18.324 [REPO] Http GET request details:
GET / HTTP/1.1

Content-Type: application/json




2022-07-03 11:14:18.456 [REPO] Response status: 200
2022-07-03 11:14:18.456 [REPO] Response details:
HTTP/1.1 200 OK

Cache-Control: max-age=0, no-cache, no-store

Connection: keep-alive

Content-Length: 853

Content-Type: application/json; charset=utf-8

Date: Sun, 03 Jul 2022 09:14:18 GMT

Expires: Sun, 03 Jul 2022 09:14:18 GMT

MS-CV: pu9EN6qIUkG7ttVP.0

Pragma: no-cache

Server: Microsoft-HTTPAPI/2.0

X-OSG-Served-By: Torus-WESTEUROPE_Legacy00000X_1.0.0.0



{"$type":"Microsoft.Marketplace.Storefront.StoreEdgeFD.BusinessLogic.Response.PackageMetadata.PackageMetadataResponse, StoreEdgeFD","Data":{"$type":"Microsoft.Marketplace.Storefront.StoreEdgeFD.BusinessLogic.Response.PackageMetadata.PackageMetadataData, StoreEdgeFD","SourceIdentifier":"StoreEdgeFD","SourceAgreements":{"$type":"Microsoft.Marketplace.Storefront.StoreEdgeFD.BusinessLogic.Response.PackageMetadata.SourceAgreements, StoreEdgeFD","AgreementsIdentifier":"StoreEdgeFD","Agreements":[{"$type":"Microsoft.Marketplace.Storefront.StoreEdgeFD.BusinessLogic.Response.PackageManifest.AgreementDetail, StoreEdgeFD","AgreementLabel":"Terms of Transaction","AgreementUrl":"https://aka.ms/microsoft-store-terms-of-transaction"}]},"ServerSupportedVersions":["1.0.0","1.1.0"],"RequiredQueryParameters":["market"],"RequiredPackageMatchFields":["market"]}}
2022-07-03 11:14:18.456 [REPO] Custom header not found.
2022-07-03 11:14:18.456 [REPO] Sending http GET request to: https://storeedgefd.dsx.mp.microsoft.com/v9.0/information
2022-07-03 11:14:18.456 [REPO] Http GET request details:
GET / HTTP/1.1

Content-Type: application/json




2022-07-03 11:14:18.494 [REPO] Response status: 200
2022-07-03 11:14:18.494 [REPO] Response details:
HTTP/1.1 200 OK

Cache-Control: max-age=0, no-cache, no-store

Connection: keep-alive

Content-Length: 853

Content-Type: application/json; charset=utf-8

Date: Sun, 03 Jul 2022 09:14:18 GMT

Expires: Sun, 03 Jul 2022 09:14:18 GMT

MS-CV: DSPzXYFBjEGRSQgy.0

Pragma: no-cache

Server: Microsoft-HTTPAPI/2.0

X-OSG-Served-By: Torus-WESTEUROPE_Legacy000005_1.0.0.0



{"$type":"Microsoft.Marketplace.Storefront.StoreEdgeFD.BusinessLogic.Response.PackageMetadata.PackageMetadataResponse, StoreEdgeFD","Data":{"$type":"Microsoft.Marketplace.Storefront.StoreEdgeFD.BusinessLogic.Response.PackageMetadata.PackageMetadataData, StoreEdgeFD","SourceIdentifier":"StoreEdgeFD","SourceAgreements":{"$type":"Microsoft.Marketplace.Storefront.StoreEdgeFD.BusinessLogic.Response.PackageMetadata.SourceAgreements, StoreEdgeFD","AgreementsIdentifier":"StoreEdgeFD","Agreements":[{"$type":"Microsoft.Marketplace.Storefront.StoreEdgeFD.BusinessLogic.Response.PackageManifest.AgreementDetail, StoreEdgeFD","AgreementLabel":"Terms of Transaction","AgreementUrl":"https://aka.ms/microsoft-store-terms-of-transaction"}]},"ServerSupportedVersions":["1.0.0","1.1.0"],"RequiredQueryParameters":["market"],"RequiredPackageMatchFields":["market"]}}
2022-07-03 11:14:18.495 [REPO] Adding to aggregated source: winget
2022-07-03 11:14:18.504 [CORE] Did not find extension: PFN = Microsoft.Winget.Source_8wekyb3d8bbwe, ID = IndexDB
2022-07-03 11:14:18.504 [REPO] Package not found Microsoft.Winget.Source_8wekyb3d8bbwe
2022-07-03 11:14:18.504 [FAIL] D:\a\_work\1\s\external\pkg\src\AppInstallerRepositoryCore\Microsoft\PreIndexedPackageSourceFactory.cpp(205)\WindowsPackageManager.dll!00007FF92EC55CE1: (caller: 00007FF92EC4B3CE) Exception(1) tid(1734) 8A15000F 
2022-07-03 11:14:18.504 [FAIL] D:\a\_work\1\s\external\pkg\src\AppInstallerRepositoryCore\RepositorySource.cpp(540)\WindowsPackageManager.dll!00007FF92ED4FD07: (caller: 00007FF92EB5BA6D) LogHr(7) tid(1734) 8A15000F     Msg:[D:\a\_work\1\s\external\pkg\src\AppInstallerRepositoryCore\Microsoft\PreIndexedPackageSourceFactory.cpp(205)\WindowsPackageManager.dll!00007FF92EC55CE1: (caller: 00007FF92EC4B3CE) Exception(1) tid(1734) 8A15000F ] 

2022-07-03 11:14:18.504 [REPO] Failed to open available source: winget
2022-07-03 11:14:18.543 [CLI ] Checking Source agreements for source: msstore
2022-07-03 11:14:18.543 [CORE] Setting action: Get, Type: Secure, Name: user_sources
2022-07-03 11:14:18.543 [CORE] Setting action: Get, Type: Standard, Name: sources_metadata
2022-07-03 11:14:18.543 [YAML] Detected UTF-8
2022-07-03 11:14:18.544 [CLI ] Source agreements satisfied. Source: msstore
2022-07-03 11:14:18.544 [CLI ] Checking Source agreements for source: winget
2022-07-03 11:14:18.544 [CORE] Setting action: Get, Type: Secure, Name: user_sources
2022-07-03 11:14:18.544 [CORE] Setting action: Get, Type: Standard, Name: sources_metadata
2022-07-03 11:14:18.544 [YAML] Detected UTF-8
2022-07-03 11:14:18.544 [CLI ] Source agreements satisfied. Source: winget
2022-07-03 11:14:18.544 [REPO] Search request meets optimized search criteria.
2022-07-03 11:14:18.547 [REPO] Sending http GET request to: https://storeedgefd.dsx.mp.microsoft.com/v9.0/packageManifests/WhatsApp.WhatsApp?Market=NL
2022-07-03 11:14:18.547 [REPO] Http GET request details:
GET / HTTP/1.1

Content-Type: application/json

Version: 1.1.0




2022-07-03 11:14:18.563 [REPO] Response status: 200
2022-07-03 11:14:18.563 [REPO] Response details:
HTTP/1.1 200 OK

Cache-Control: max-age=0, no-cache, no-store

Connection: keep-alive

Content-Length: 132

Content-Type: application/json; charset=utf-8

Date: Sun, 03 Jul 2022 09:14:18 GMT

Expires: Sun, 03 Jul 2022 09:14:18 GMT

MS-CV: 4b3qXBn6tE6OFlVK.0

Pragma: no-cache

Server: Microsoft-HTTPAPI/2.0

X-OSG-Served-By: Torus-WESTEUROPE_Legacy000021_1.0.0.0



{"$type":"Microsoft.Marketplace.Storefront.StoreEdgeFD.BusinessLogic.Response.PackageManifest.PackageManifestResponse, StoreEdgeFD"}
2022-07-03 11:14:18.563 [REPO] No manifest results returned.
2022-07-03 11:14:18.564 [CLI ] Caught wil::ResultException: D:\a\_work\1\s\external\pkg\src\AppInstallerRepositoryCore\Microsoft\PreIndexedPackageSourceFactory.cpp(205)\WindowsPackageManager.dll!00007FF92EC55CE1: (caller: 00007FF92EC4B3CE) Exception(1) tid(1734) 8A15000F 
```

I think these are the most important bits:

``` text
2022-07-03 11:14:15.765 [REPO] Source past auto update time [5 mins]; it has been at least 27613994 mins
2022-07-03 11:14:16.089 [FAIL] WindowsPackageManager.dll!00007FF92ED457C2: ReturnHr(1) tid(1734) 80072EFD     Msg:[winrt::hresult_error: Kan geen verbinding met de server maken.] 

2022-07-03 11:14:16.089 [FAIL] WindowsPackageManager.dll!00007FF92EBF59D7: LogHr(2) tid(1734) 80072EFD 
2022-07-03 11:14:16.090 [FAIL] D:\a\_work\1\s\external\pkg\src\AppInstallerRepositoryCore\RepositorySource.cpp(53)\WindowsPackageManager.dll!00007FF92ED503A1: (caller: 00007FF92EC4AEA6) LogHr(3) tid(1734) 80072EFD     Msg:[winrt::hresult_error: Kan geen verbinding met de server maken.] 

2022-07-03 11:14:16.090 [REPO] Source add/update failed, waiting a bit and retrying: winget
2022-07-03 11:14:18.323 [FAIL] WindowsPackageManager.dll!00007FF92ED457C2: ReturnHr(2) tid(1734) 80072EFD     Msg:[winrt::hresult_error: Kan geen verbinding met de server maken.] 

2022-07-03 11:14:18.323 [FAIL] WindowsPackageManager.dll!00007FF92EBF59D7: LogHr(5) tid(1734) 80072EFD 
2022-07-03 11:14:18.323 [FAIL] D:\a\_work\1\s\external\pkg\src\AppInstallerRepositoryCore\RepositorySource.cpp(516)\WindowsPackageManager.dll!00007FF92ED4FB13: (caller: 00007FF92EB5BA6D) LogHr(6) tid(1734) 80072EFD     Msg:[winrt::hresult_error: Kan geen verbinding met de server maken.] 

2022-07-03 11:14:18.323 [REPO] Failed to update source: winget
```

and

``` text
2022-07-03 11:14:18.504 [REPO] Package not found Microsoft.Winget.Source_8wekyb3d8bbwe
2022-07-03 11:14:18.504 [FAIL] D:\a\_work\1\s\external\pkg\src\AppInstallerRepositoryCore\Microsoft\PreIndexedPackageSourceFactory.cpp(205)\WindowsPackageManager.dll!00007FF92EC55CE1: (caller: 00007FF92EC4B3CE) Exception(1) tid(1734) 8A15000F 
2022-07-03 11:14:18.504 [FAIL] D:\a\_work\1\s\external\pkg\src\AppInstallerRepositoryCore\RepositorySource.cpp(540)\WindowsPackageManager.dll!00007FF92ED4FD07: (caller: 00007FF92EB5BA6D) LogHr(7) tid(1734) 8A15000F     Msg:[D:\a\_work\1\s\external\pkg\src\AppInstallerRepositoryCore\Microsoft\PreIndexedPackageSourceFactory.cpp(205)\WindowsPackageManager.dll!00007FF92EC55CE1: (caller: 00007FF92EC4B3CE) Exception(1) tid(1734) 8A15000F ] 

2022-07-03 11:14:18.504 [REPO] Failed to open available source: winget
```

Since a ["80072EFD" "winget" Google Search](https://www.google.com/search?q=%2280072EFD%22+%22winget%22) returns remarkably few results (and just a pr0n link on ["80072EFD" "WindowsPackageManager.dll"](https://www.google.com/search?q=%2280072EFD%22+%22WindowsPackageManager.dll%22)), I need help figuring out what `WindowsPackageManager.dll` is trying to access.

Anyone having hints or tips for this?

(Maybe I Googled for the wrong search phrase: any help with that is appreciated too; aged 53 I still love learning new things every day, even things that sound obvious to others).

Thanks in advance,

--jeroen

**Comment by jpluimers** [comment, marked as answer]
Did quite a bit of fiddling, deduction and searching based on the log file (more on that in a later blog post on [wiert.me](https://wiert.me)). This is the conclusion:

If I had done ["winget" "0x8a15000f" - Google Search](https://www.google.com/search?q=%22winget%22+%220x8a15000f%22) instead of ["80072EFD" "winget" Google Search](https://www.google.com/search?q=%2280072EFD%22+%22winget%22), then I would have found [0x8a15000f : Data required by the source is missing · Issue #1348 · microsoft/winget-cli](https://github.com/microsoft/winget-cli/issues/1348):

> The output of the command `winget source list`
> 
> ``` text
> Name    Argument
> -----------------------------------------------------
> winget  https://winget.azureedge.net/cache
> ```

Adding `winget.azureedge.net` to the Fritz!Box router whitelist solved the `0x8a15000f : Data required by the source is missing`.

Things like these (both the failing `https://winget.azureedge.net/cache/source.msix` URL and the `0x8a15000f ` error code) should be in the logs, so I will add a new issue for that later.

Note that the now failing WhatsApp install URL *is* in the log, so that was sort of easy to fix when following the redirect (see the [archived redirect checker result](https://web.archive.org/web/20220708165653/https://www.whatsmydns.net/redirect-checker?q=https%3A%2F%2Fweb.whatsapp.com%2Fdesktop%2Fwindows%2Frelease%2Fx64%2FWhatsAppSetup.exe)).

Basically `https://web.whatsapp.com/desktop/windows/release/x64/WhatsAppSetup.exe` (which is in the log) redirects to somewhere on the [FBCDN](https://en.wikipedia.org/wiki/FBCDN.Net) (which is not in the log) so I to whitelist `fbcdn.net` completely.

--jeroen
